pkg/auth/authenticator.go

```diff
@@ -261,3 +271,52 @@ func CheckDomain(domain string, domainAllowedList []string) (bool, error) {
 	}
 	return false, nil
 }
+
+func (a *Authenticator) checkServiceUpstreamAuth(wsvc string, request *Request, extraHeaders *ExtraHeaders) (bool, CerberusReason) {
+	service, ok := (*a.servicesCache)[wsvc]
+	if !ok {
+		return false, CerberusReasonWebserviceNotFound
+	}
+	if service.Spec.UpstreamHttpAuth == (cerberusv1alpha1.UpstreamHttpAuthService{}) {
+		return true, CerberusReasonOK
+	}
+	if service.Spec.UpstreamHttpAuth.ReadTokenFrom == "" {
+		return false, CerberusReasonSourceAuthTokenEmpty
+	}
+	if service.Spec.UpstreamHttpAuth.WriteTokenTo == "" {
+		return false, CerberusReasonTargetAuthTokenEmpty
+	}
+	if !govalidator.IsRequestURL(service.Spec.UpstreamHttpAuth.Address) {
+		return false, CerberusReasonInvalidUpstreamAddress
+	}
+
+	token := request.Request.Header.Get(service.Spec.UpstreamHttpAuth.ReadTokenFrom)
+
+	// TODO: get http method from webservice crd
+	req, err := http.NewRequest("GET", service.Spec.UpstreamHttpAuth.Address, nil)
+	if err != nil {
+		return false, CerberusReasonUpstreamAuthFailed
+	}
+
+	req.Header = http.Header{
+		service.Spec.UpstreamHttpAuth.WriteTokenTo: {token},
+		"Content-Type": {"application/json"},
+	}
+
+	resp, err := a.httpClient.Do(req)
+	if err != nil {
+		return false, CerberusReasonUpstreamAuthFailed
+	}
+	if resp.StatusCode != http.StatusOK {
+		return false, CerberusReasonUnauthorized
+	}
+	var headersString string
+	for header, values := range resp.Header {
+		for _, value := range values {
+			headersString += header + ": " + value + "\n"
```

Comment:
why are we concatenating  ```\n``` to headers? and I think we should add ```CareHeaders``` to ```HttpUpstreamAuth``` to tell Cerberus which headers from upstream should be added to downstream response.

Reply:
GOOD IDEA,
careHeaders has been added too.